### DISCUSSION TITLE
Dual application with zephyr mcuboot for stm32h750b_dk

### DISCUSSION THREAD

**Discussion by MaazSk** [zephyrproject-rtos/zephyr#79688] [Q&A]
I am using stm32h750b_dk, I have mcuboot I want it to boot from different partition where I have my two different application stored. 

- Can you please help me with the step to step guide how this can be achieved? 
- Is there in example provided by zephyr which I can look into?
- is it possible through west flash or any other tools are required?

**Comment by erwango** [comment]
@FRASTM maybe ?

**Comment by JarmouniA** [comment]
@MaazSk This may be helpful https://github.com/zephyrproject-rtos/zephyr/discussions/73570

**Comment by FRASTM** [comment]
@MaazSk 
If you do not use the external NOR flash, I am not sure this is simple: the pb I see is that the stm32h750 has only 1 sector of 128K
Even if you set partition with mcuboot + application 1 + application 1  (which is tiny), the download of flash (west flash) is erasing the sector before writing, so each time you write the flash all the content is erased.

 However the principle is to
- define the flash memory partitionning : 
  mcuboot at offset 0  (zephyr,code-partition = &boot_partition;  in the board overlay inside the ..//bootloader/mcuboot/boot/zephyr/
  image -0 at slot0   at offset 0x10000 (assuming the mcu boot partition is 64KB) (zephyr,code-partition = slot0_partition in the board overlay inside the ./zephyr/ )
  image-1 at slot1 at offset 0x18000  (assuming your applications < 32KB)   (zephyr,code-partition = slot1_partition in the board overlay inside the ./zephyr/ )
  
- build and link both applications to run at their own flash address
 build  with --sysbuild  -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y

- flash the mcu_boot and one code at slot0 and the other at slot1
- run from mcuboot to choose the image you want to execute.

Another possible solution could be to use one flash bank per application and to swap banks  (no mcuboot involved)

Then if you need more flash space, you probably need to enable the external NOR in MemoryMapped mode.

**Reply by MaazSk** [reply]
Hi @FRASTM,

Thank you for the reply. I am using the on board external flash which is mt25ql512ab1 and I have made a change to dts as follows

```
mt25ql512ab1: qspi-nor-flash-1@90000000 {
		compatible = "st,stm32-qspi-nor";
		reg = <0x90000000 DT_SIZE_M(64)>; /* 512 Mbits */
		qspi-max-frequency = <72000000>;
		spi-bus-width = <4>;
		reset-cmd;
		reset-cmd-wait = <20>;
		status = "okay";

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			slot0_partition: partition@0 {
				label = "image-0";
				reg = <0x00000000 0x1000000>;
			};

			slot1_partition: partition@1000000{
				label = "image-1";
				reg = <0x1000000 DT_SIZE_M(64)>;
			};
		};
	};
```
I am able to store to different location using this in my application overlay file

```
/ {
	chosen {
		zephyr,code-partition = &slot0_partition;
	};
};
```

So changing the `slot0_partition` to `slot1_partition` it flash to different slot and able to boot from there.

Now here is the thing which I want to achieve
When there is a button press I want to switch from one application to another application
I am not sure how to achieve this where should I make the changes.

**Comment by FRASTM** [comment]
This is interesting. First of all I guess you can execute application 1 or application 2 on the external flash in XiP mode.
I did not explore that kind of behavior : selecting application 1 or application 2 with a button.
I would recommend to have a look around the [mcumgr](https://docs.zephyrproject.org/latest/services/device_mgmt/mcumgr.html) which has the capability to choose one image on a flash partition. Another possibility would be to directly hack the mcuboot drives the user push-button and boot on the selected image.

**Reply by MaazSk** [reply]
Looking into it straight away. 
Yes, currently I can only run one application at a time and my `west flash` erases the complete flash so, need a way to erase only particular sector.

Thank you! for this I will update soon as I make progress

**Reply by FRASTM** [reply]
> Looking into it straight away. Yes, currently I can only run one application at a time and my `west flash` erases the complete flash so, need a way to erase only particular sector.
> 
> Thank you! for this I will update soon as I make progress

If you "west flash" the stm32h750b disco as there is only one 128KB sector, it is erased. That's why you need to place the boot loader in the internal flash (when `zephyr,code-partition = &boot_partition;` and with west flash) and download _manually_ the application 1 & 2 in the external flash image 0 & 1: this is done with the stm32Cube programmer selecting the external loader MT25TL01G_STM32H750B-DISCO.stldr  for that disco board.
This is something I experienced around in https://github.com/zephyrproject-rtos/zephyr/pull/64641 and https://github.com/zephyrproject-rtos/mcuboot/pull/112

**Reply by MaazSk** [reply]
okay, so this is the missing block I think

**Reply by MaazSk** [reply]
Hi @FRASTM,

I was trying out the mcumgr but was unable to make it work and I think uart is not responding could you please check my discussion [over here](https://github.com/zephyrproject-rtos/zephyr/discussions/81421)
